Fix a bug where a recursively added `EndpointGroup` listener is not invoked

ikhoon opened the pull request:
Motivation:

`WeightedRoundRobinStrategy` fails to select an `Endpoint` from a non-empty `EndpointGroup`.
```
c.l.a.c.UnprocessedRequestException: c.l.a.c.e.EndpointSelectionTimeoutException:
  Failed to select within 5000 ms an endpoint from: DnsAddressEndpointGroup{selectionStrategy=class
  com.linecorp.armeria.client.endpoint.WeightedRoundRobinStrategy, allowsEmptyEndpoints=false, initialized=true, numEndpoints=1,
  endpoints=[Endpoint{xxx.com:8443, ipAddr=xxx.xxx.xxx.xxx, weight=1000}], questions=[DnsQuestion(xxx.com. IN A)], logPrefix=xxx.com., attemptsSoFar=0}
	at com.linecorp.armeria.client.UnprocessedRequestException.of(UnprocessedRequestException.java:45)
	at com.linecorp.armeria.client.HttpClientDelegate.earlyFailedResponse(HttpClientDelegate.java:228)
	at com.linecorp.armeria.client.HttpClientDelegate.execute(HttpClientDelegate.java:66)
	at com.linecorp.armeria.client.HttpClientDelegate.execute(HttpClientDelegate.java:51)
```

Through static analysis, I found out that a race condition exists in our `Endpoint` selection logic.

1. `ListeningFuture` receives a notification of updated endpoints and attempts to select endpoints. https://github.com/line/armeria/blob/d5f65e5a3703282fe28b8336628079a1d32f161d/core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java#L139
2. They are the initial endpoints. As `EndpointSelector` is lazily created, `DynamicEndpointGroup` creates the EndpointSelector first. https://github.com/line/armeria/blob/d5f65e5a3703282fe28b8336628079a1d32f161d/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java#L195
3. In `EndpointSelector` constructor, it adds a new listener in order to get the current endpoints. The new listener is added while the original listener is invoked. https://github.com/line/armeria/blob/d5f65e5a3703282fe28b8336628079a1d32f161d/core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java#L59-L62
4. After all listeners are called, `initialEndpointsFuture` is completed. Thus the listener added in 3) can't immediately be called despite the presence of current endpoints. Because `latestValue()` returns the endpoints when `initialEndpointsFuture` is completed. https://github.com/line/armeria/blob/d5f65e5a3703282fe28b8336628079a1d32f161d/core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java#L226-L227 https://github.com/line/armeria/blob/863e27c5268631360c14e50e13cd3e38d84c9db5/core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointGroup.java#L27-L28
5. As a result of 4), endpointsAndWeights remains null, causing `selectNow(ctx)` to return null. https://github.com/line/armeria/blob/d5f65e5a3703282fe28b8336628079a1d32f161d/core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java#L67-L70

Modifications:

- Make `AbstractEndpointSelector` implement `Consumer<List<Endpoint>>` to listener to new endpoints itself.
  - When new endpoints are updated, it is propagated to subclasses via `updateNewEndpoints()`.
  - If no endpoints are available, `select()` adds an incomplete `ListeningFuture` to `pendingFutures` and return the `ListeningFuture`.
  - If new endpoints are updated:
    - `updateNewEndpoints()` is invoked to initialize or update `EndpointSelectionStrategy`.
    - The `ListeningFuture` completes and it is removed from `pendingFutures`.
- Fixed `EndpointSelectionStrategy` implementations not to use endpoint listeners.
  - When using multiple listeners, there were dependencies between listeners.
    - That has caused many concurrency issues, e.g. #4448, #3978, the bug in the motivation and so on.
    - New bugs can easily be made even with small modifications.
- Complete `DynamicEndpoint.whenReady()` before notifying listeners so that the listeners can access the latest values via `latest()`.
- Notify a new listener to which `notifyLatestValue` is set after adding the listener to `updateListeners`. Otherwise, `removeListener()` invoked the listener is ignored and leaked.
- Miscellaneous) Optimize `latestValue()` of `DynamicEndpointGroup`.


Result:

You no longer see a false negative `EndpointSelectionTimeoutException` when `EndpointSelectionStrategy.weightedRoundRobin()` is used with `DynamicEndpointGroup`.

minwoox reviewed (approved):
Thanks! 👍

jrhee17 commented:
Is this PR ready for review? 😄    Let me know if it is ready 🙏

ikhoon commented:
Not yet. I will address @minwoox's idea that use the lock in the `AbstractListener`.

ikhoon commented:
Addressed all comments. PTAL.

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:
```diff
@@ -213,37 +210,39 @@ public final CompletableFuture<List<Endpoint>> whenReady() {
-        notifyListeners(newEndpoints);
-        completeInitialEndpointsFuture(newEndpoints);
+    @Override
+    protected final List<Endpoint> latestValue() {
```
Question) is this override still necessary?

From what I understood, if `setEndpoints` is called before or after `endpointGroup#addListener` then we shouldn't have a race condition anymore.

https://github.com/line/armeria/blob/1c93feef523b619d1d73109772fe1530fb0f110e/core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java#L76-L76

ikhoon replied:
Right. Technically, we don't have to override it since the race condition has been gone. However, I didn't revert it since the new version is slightly optimized.

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:
```diff
@@ -260,15 +259,18 @@ protected final void setEndpoints(Iterable<Endpoint> endpoints) {
         }
 
-        endpointsLock.lock();
+        lock();
```
Question) Would it be worth acquiring a lock before L258 so that `hasChanges` doesn't return early unintentionally?

ikhoon replied:
Good suggestion. Fixed.

jrhee17 reviewed (commented):
Left some minor questions, but basically changes look good 👍

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:
```diff
@@ -253,22 +252,26 @@ protected final void setEndpoints(Iterable<Endpoint> endpoints) {
 
+    private void unsafeUpdateEndpoints(List<Endpoint> newEndpoints) {
+        endpoints = newEndpoints;
         notifyListeners(newEndpoints);
```
Shouldn't we call `listener.accept(latestValue);` out of the lock?
How about adding `getListeners` to `AbstractListenable` and calling
```java
for (Consumer<? super T> listener : listeners) {
    listener.accept(latestValue);
}
```
out of the lock as is?

ikhoon replied:
I'd tried to refactor the code to avoid locking for notification but the code becomes too complex.
```java
protected final void removeEndpoint(Endpoint e) {
    lock();
    boolean unlocked = false;
    try {
        final List<Endpoint> newEndpoints = endpoints.stream()
                                                     .filter(endpoint -> !endpoint.equals(e))
                                                     .collect(toImmutableList());
        updateEndpoints(newEndpoints, true);
        unlocked = true;
    } finally {
        if (!unlocked) {
            unlock();
        }
    }
}

private void updateEndpoints(Iterable<Endpoint> endpoints, boolean locked) {
    if (!locked) {
        lock();
    }

    List<Consumer<List<Endpoint>>> listeners = null;
    List<Endpoint> newEndpoints = null;
    try {
        if (!allowEmptyEndpoints && Iterables.isEmpty(endpoints)) {
            return;
        }

        final List<Endpoint> oldEndpoints = this.endpoints;
        final List<Endpoint> sorted = ImmutableList.sortedCopyOf(endpoints);
        if (!hasChanges(oldEndpoints, sorted)) {
            return;
        }

        this.endpoints = newEndpoints = sorted;
        listeners = listeners();
    } finally {
        unlock();

        if (newEndpoints != null) {
            if (listeners != null) {
                for (Consumer<List<Endpoint>> listener : listeners) {
                    listener.accept(newEndpoints);
                }
            }
            maybeCompleteInitialEndpointsFuture(newEndpoints);
        }
    }
}
```

Are you worried about if a listener acquires a lock for a long time?

minwoox replied:
> Are you worried about if a listener acquires a lock for a long time?

Yes, and I think the current design is based on the same notion. We don't usually call listeners while holding a lock.
But if it's unavoidable, I think we can make an exception for the case.

> t the code becomes too complex.

I was rather thinking something like this:
```java
protected final void setEndpoints(Iterable<Endpoint> endpoints) {
    if (!allowEmptyEndpoints && Iterables.isEmpty(endpoints)) {
        return;
    }

    // Might need to change the type if it's ugly.
    final Consumer<? super List<Endpoint>>[] listeners;
    final List<Endpoint> newEndpoints;
    lock();
    try {
        final List<Endpoint> oldEndpoints = this.endpoints;
        newEndpoints = ImmutableList.sortedCopyOf(endpoints);

        if (!hasChanges(oldEndpoints, newEndpoints)) {
            return;
        }

        this.endpoints = newEndpoints;
        listeners = copyListeners();
    } finally {
        unlock();
    }

    for (Consumer<? super List<Endpoint>> listener : listeners) {
        listener.accept(newEndpoints);
    }
    maybeCompleteInitialEndpointsFuture(newEndpoints);
}
```

ikhoon replied:
> We don't usually call listeners while holding a lock.

Q: Should we fix the following code block as well?
https://github.com/line/armeria/blob/496a3f63621b68b73fbd626878c811c2acf3d007/core/src/main/java/com/linecorp/armeria/common/util/AbstractListenable.java#L90-L95

minwoox commented:
Please let me know when this is ready. 😉

ikhoon commented:
Sorry, I missed the test failure. `notifyListeners` invoked `ListeningFuture.accept()` first and then invoked the listener added by `WeightedRoundRobinSelector`.
https://github.com/line/armeria/blob/fb04123b58a2ccba5a33def5d00753f53978048c/core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java#L133-L133
https://github.com/line/armeria/blob/43a4ae496db6956a01a3ab70f2f62c2cb447582f/core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java#L61
When `selectNow()` was called in `ListeningFuture.accept()` returned null because `WeightedRoundRobinSelector` had not been initialized.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java:
```diff
@@ -132,29 +132,32 @@ final class RampingUpEndpointWeightSelector extends AbstractEndpointSelector {
             super(endpointGroup);
             this.executor = executor;
 
-            final AtomicBoolean initialized = new AtomicBoolean();
```
Fixed due to https://github.com/line/armeria/pull/5163#issuecomment-1770436611

trustin reviewed (approved):
:+1:

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/DynamicEndpointGroup.java:
```diff
@@ -291,7 +290,17 @@ private static boolean hasChanges(List<Endpoint> oldEndpoints, List<Endpoint> ne
 
-    private void completeInitialEndpointsFuture(List<Endpoint> endpoints) {
+    @Override
+    protected List<Endpoint> latestValue() {
```
Do we need this method now?

ikhoon replied:
Not for correctness. It was overridden to provide an optimized implementation.

minwoox replied:
Isn't caliing ` whenReady().isDone()` is more optimized than accessing the volatile `endpoints` field when it's not initialized?

ikhoon replied:
If `isDone()` is true, it would access two volatile fields.

minwoox replied:
Ah, that is true. 👍

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java:
```diff
@@ -132,29 +132,32 @@ final class RampingUpEndpointWeightSelector extends AbstractEndpointSelector {
-                }
-            }, true);
-
+            endpointGroup.whenReady().thenAccept(unused -> maybeUpdateEndpoints(endpointGroup.endpoints()));
```
How about
```suggestion
            endpointGroup.whenReady().thenAccept(this::maybeUpdateEndpoints);
```
so that we don't access the volatile field?

ikhoon replied:
`endpointGroup.whenReady()` completes with the initial value. If this callback is added after updating new endpoints which are never updated, the listener is not invoked.
`endpointGroup.endpoints()` is used to acquire the latest value instead of the initial value.

minwoox replied:
> If this callback is added after updating new endpoints 

If so the callback is invoked right away. Isn't it?

ikhoon replied:
If the callback is added after the last update, it wouldn't be invoked.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java:
```diff
@@ -132,29 +132,32 @@ final class RampingUpEndpointWeightSelector extends AbstractEndpointSelector {
+                endpointSelector = new WeightedRandomDistributionEndpointSelector(dedupEndpoints);
+                endpointsFinishedRampingUp.addAll(dedupEndpoints);
+
+                group().addListener(this::maybeUpdateEndpoints);
```
Can't this be
```suggestion
                group().addListener(this::updateEndpoints);
```
because `initialized` never changes back to `false`?

ikhoon replied:
Oops... that's my fault.

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -58,8 +58,17 @@ private static final class WeightedRoundRobinSelector extends AbstractEndpointSe
             super(endpointGroup);
-            endpointGroup.addListener(endpoints -> endpointsAndWeights = new EndpointsAndWeights(endpoints),
-                                      true);
+            endpointGroup.whenReady().thenAccept(unused -> maybeUpdateEndpoints(endpointGroup.endpoints()));
```
ditto

ikhoon replied:
Commented at https://github.com/line/armeria/pull/5163#discussion_r1370052712

minwoox reviewed (commented):
Looks good overall. 👍 
Left some suggestions. 😉

minwoox commented on core/src/main/java/com/linecorp/armeria/common/util/AbstractListenable.java:
```diff
@@ -35,30 +34,19 @@
-
-    private final ReentrantLock reentrantLock = new ReentrantShortLock();
+    // Updated via updateListenersUpdater
+    @SuppressWarnings("unchecked")
```
This isn't addressed?

ikhoon replied:
Fixed.

minwoox reviewed (approved):
👍 👍 👍

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java:
```diff
@@ -132,29 +131,31 @@ final class RampingUpEndpointWeightSelector extends AbstractEndpointSelector {
         }
 
+        private void initializeEndpointSelector(List<Endpoint> endpoints) {
+            final List<Endpoint> dedupEndpoints =
```
How about adding `currentEndpoints = endpoints;`

minwoox commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightRampingUpStrategy.java:
```diff
@@ -262,7 +267,7 @@ private Set<EndpointAndStep> filterOldEndpoints(List<Endpoint> newEndpoints) {
 
             final List<Endpoint> replacedEndpoints = new ArrayList<>();
-            for (final Iterator<Endpoint> i = endpointsFinishedRampingUp.iterator(); i.hasNext();) {
+            for (final Iterator<Endpoint> i = endpointsFinishedRampingUp.iterator(); i.hasNext(); ) {
```
Could you lint again?

jrhee17 commented on core/src/main/java/com/linecorp/armeria/common/util/AbstractListenable.java:
```diff
@@ -85,28 +72,40 @@ public final void addListener(Consumer<? super T> listener) {
-                if (latest != null) {
-                    listener.accept(latest);
-                }
+        if (notifyLatestValue) {
```
Question) I'm wondering if CAS is really safe in this case since we may call `removeListener` from listeners like done in `AbstractEndpointSelector`

https://github.com/line/armeria/blob/9ddd6c1b76fdc0fe8943dd417e80773cb92eebdc/core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java#L141

```java
static class MyListener implements Consumer<List<Endpoint>> {

    final EndpointGroup group;
    int invoked = 0;
    MyListener(EndpointGroup group) {
        this.group = group;
    }

    @Override
    public void accept(List<Endpoint> endpoints) {
        invoked++;
        group.removeListener(this);
    }

    public int invoked() {
        return invoked;
    }
}

@Test
void shouldNotifyListenerWithLatestValue2() throws InterruptedException {
    final DynamicEndpointGroup group = new DynamicEndpointGroup();
    group.setEndpoints(ImmutableList.of(Endpoint.of("foo")));
    final MyListener listener = new MyListener(group);
    group.addListener(listener, true);
    assertThat(listener.invoked).isEqualTo(1);
    // listener isn't removed at this point, possible leak
}
```

ikhoon replied:
Good point. Would it be safe after the CAS operation for adding is done?
https://github.com/line/armeria/blob/a2be9d95e16319a3d4ebe5928bdf9dd6c4fa7422/core/src/main/java/com/linecorp/armeria/common/util/AbstractListenable.java#L85-L85

jrhee17 replied:
I guess so.

I'm also a little wary that the count of `listeners` can be high and cause some contention (since `listeners` is bound by the number of requests which can be large for all we know)

Was this change due to a bug? or was it a refactoring?

ikhoon replied:
> since listeners is bound by the number of requests which can be large for all we know

In normal cases, the number of listeners is not bound by requests. `ListeningFuture` will only be used when an `EndpointGroup` is not initialized or becomes empty because of an outage. 

It was changed to avoid invoking external callbacks with the lock which could cause potential deadlock.

ikhoon replied:
After looking at AbstractEndpointSelector, I don't think we need to add a new listener for every `select()` method for notification. Let me tweak it a bit.

jrhee17 replied:
I see, we're not doing any scheduling to other threads inside the lock so I didn't think there was a possibility of deadlock.
I guess it's fine to leave as is then.

ikhoon replied:
Fixed. (Sorry for the frequent changes. 😅)

jrhee17 commented on core/src/main/java/com/linecorp/armeria/client/endpoint/WeightedRoundRobinStrategy.java:
```diff
@@ -58,8 +58,17 @@ private static final class WeightedRoundRobinSelector extends AbstractEndpointSe
             super(endpointGroup);
-            endpointGroup.addListener(endpoints -> endpointsAndWeights = new EndpointsAndWeights(endpoints),
-                                      true);
+            endpointGroup.whenReady().thenAccept(unused -> maybeUpdateEndpoints(endpointGroup.endpoints()));
```
Minor question:
Was this also an optimization? or is this change necessary? (It's hard to track which changes correspond to which motivation 😅 )

I prefer that we always try to use listeners instead of `whenReady` since:
1) Users usually call `whenReady.join()` on application startup. If listeners are invoked beforehand there is less probability of race conditions
2) Less complexity - we only have to care about a single callback. I feel like most of this confusion is because we are mixing usages of `initialCompleteFuture` with `listener`s

```suggestion
            endpointGroup.addListener(endpoints -> this.endpointsAndWeights = new EndpointsAndWeights(endpoints));
```

ikhoon replied:
I removed the whole code but let me explain my original purpose. I changed `AbstractEndpointSelector()` to add an endpoint listener first when it is created which resulted in adding the listener of `WeightedRoundRobinSelector` later. Now we have two endpoint listeners.

If the `EndpointGroup` is not initialized yet, `AbstractEndpointSelector.select()` creates `ListeningFuture` and waits for new endpoints via the listener. When `AbstractEndpointSelector` receives the event, it consequently calls `WeightedRoundRobinSelector.selectNow()`. At the moment, the listener of `WeightedRoundRobinSelector` is not invoked yet and its `SelectionSterategy` will be empty or null.

For that reason, I used `whenReady()` to initialize `SelectionSterategy` before `AbstractEndpointSelector` receives an endpoint update event.

jrhee17 reviewed (approved):
Thanks for handling this issue @ikhoon 🙇 👍 🙇

ikhoon commented:
`AbstractEndpointSelector` and some `EndpointSelectionStrategy` implementations depend on `EndpointGroup.addListener()`. The order of adding listeners is important. `EndpointSelectionStrategy` should add endpoint listeners before `AbstractEndpointSelector` to preliminary initialize the `EndpointSelectionStrategy`. 

If `AbstractEndpointSelector` adds an endpoint listener before `EndpointSelectionStrategy`, `EndpointSelectionStrategy` is not initialized because the listener is not invoked yet. Consequently `selectNow()` may return null although the `EndpointGroup` is initialized.

The implementations of selector and strategy are loosely coupled. It is difficult to control the order of listeners. Whenever I made a small change in the code, the endpoint selection was easily broken and added another workaround for that. As the number of workarounds increases, the direction of revision becomes further away from the initial goal of PR. 

I will think again about whether continuing to use `addListener` is the right direction. Let me reschedule this PR to 1.27.0

ikhoon commented:
New updates:
- Reverted most changes in `AbstractListenable` except the possible listener leak. https://github.com/line/armeria/pull/5163#discussion_r1371232199
- Fixed `EndpointSelectionStrategy` implementations not to use endpoint listeners.
  - New endpoints are updated via `updateNewEnpoints()` method directly by `AbstractEndpointSelector`.

ikhoon commented on core/src/test/java/com/linecorp/armeria/client/ClientRequestContextDelayedInitTest.java:
```diff
@@ -79,19 +80,28 @@ void timeout() {
      */
     @ParameterizedTest
-    @CsvSource({ "0", "1", "2", "3" })
+    @CsvSource({ "0", "1", "2" })
```
one `selectNow()` has been removed so the maximum number of invocation of `selectNow()` is 3.

ikhoon replied:
https://github.com/line/armeria/pull/5163/files#diff-c61e716447f997dea21163d52d7a0a59fc77c25f1cab14426875c7b109876800L80

ikhoon commented:
This PR is ready to review. All tests are passed locally.

ikhoon commented on core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java:
```diff
@@ -115,8 +124,59 @@ public final CompletableFuture<Endpoint> select(ClientRequestContext ctx,
+     * {@link EndpointGroup}. The new endpoints will be passed to {@link #updateNewEndpoints(List)}.
+     */
+    @UnstableApi
+    protected final void initialize() {
```
This method was split from the constructor because if the `EndpointGroup` is initialized already, `updateNewEndpoints()` can access member fields from the subclass while initializing the class.

trustin commented on core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java:
```diff
@@ -26,21 +28,31 @@
  * the change events emitted by {@link EndpointGroup} specified at construction time.
  */
-public abstract class AbstractEndpointSelector implements EndpointSelector {
+public abstract class AbstractEndpointSelector implements EndpointSelector, Consumer<List<Endpoint>> {
```
Could you define a listener field instead of implementing it? i.e.

```java
public abstract class AbstractEndpointSelector implements EndpointSelector {
    private final Consumer<List<Endpoint>> listener = endpoints -> {
        updateNewEndpoints(endpoints);
        ...
    }
}
```

ikhoon replied:
Sounds good. Method reference is used as a listener instead.

trustin reviewed (approved):
:+1:

minwoox reviewed (approved):
Looks great! 👍 👍 👍

minwoox commented:
Thanks again, @ikhoon! 🚀 🚀 🚀
